### DISCUSSION TITLE
Using Lowdefy for non-internal tools?

### DISCUSSION THREAD

**Discussion by carwei** [lowdefy/lowdefy#1409] [General]
Just curious. Is there any specific reason the tagline for Lowdefy is "Create **internal** tools with YAML" (my emphasis)? This seems promising to use in external applications, eg. rapidly developing a SaaS.

**Comment by carwei** [comment]
One thing I realized is that there is no server-side validation built in. It all depends on the use case and risk of users bypassing your validation, but this is potentially more of a drawback for something deployed to external users. Anything else?

**Comment by Gervwyk** [comment]
Hi @carwei 

We decided to launch Lowdefy as an internal tool builder and later transition to any web app. The reasoning behind this is that we primarily use Lowdefy to build internal tools for customers. This means Lowdefy is excellent at building data intensive web apps, however not currently designed to build pixel perfect apps (custom styling), which is more important in consumer facing app - we plan to improve on this in the future and also serve this requirement well, v4 already greatly improves on this.

Regarding server-side validation. You are right. Currently, you can connect to a custom API if that is a requirement. However, we have some exciting ideas on how this will be implemented in Lowdefy in the future.

So in short. Yes. You can build any web app with Lowdefy. One day soon, our marketing will also say that.

**Reply by carwei** [reply]
Thank you @Gervwyk, this makes a lot of sense. It's so much easier to focus on the clients you currently serve compared to guessing what someone else might need. I don't have a pixel-perfect requirement, but looking forward to seeing Lowdefy evolve.

> Regarding server-side validation. You are right. Currently, you can connect to a custom API if that is a requirement. However, we have some exciting ideas on how this will be implemented in Lowdefy in the future.

Could you give me a rough idea of how far into the future this is (more or less than a year)? Just to get a sense of how much of a workaround is reasonable to invest in for now.

I think that the solution right now would be to have Lowdefy serve the front-end and set up a back-end in a separate Node process to serve an API that includes server-side validation (one big drawback that comes to mind is that I need to set up authorization and authentication separately for the back-end to mirror the front-end roles).

**Reply by Gervwyk** [reply]
>  think that the solution right now would be to have Lowdefy serve the front-end and set up a back-end in a separate Node process to serve an API that includes server-side validation (one big drawback that comes to mind is that I need to set up authorization and authentication separately for the back-end to mirror the front-end roles).

So with v4 this is not an issue since you can write your own connectors / requests easily as a plugin. The requests runs after all auth checks are completed, so they are just as secure as the default requests.

I'll see if I can get a public example repo up of set for custom plugins, since currently we still need to update the docs for this.
I think this will be the best approach, as I cannot provide a timeline currently for the Lowdefy backend request pipeline features, also the custom request approach gives you all the flexibility of code. So this is a great option!

**Reply by carwei** [reply]
Ok, that would be great. Do custom plugins only work in v4? 
Is there a timeline for when v4 is considered stable? 
Thanks!

**Reply by Gervwyk** [reply]
Yes, only v4. 
It is stable, currently switching over the last of our apps. So we've been running it in prod for about 3 months now. 
It is still in alpha, since we need to document the changes and one or two things left before it has v3 feature parity. We hope to have v4 full release by end Jan.

**Reply by Gervwyk** [reply]
Here is an example of the plugins:
https://github.com/lowdefy/lowdefy-example-plugins

**Reply by carwei** [reply]
> Here is an example of the plugins: https://github.com/lowdefy/lowdefy-example-plugins

Great, thank you! And this is relevant for V3 but expanded in V4?

**Reply by Gervwyk** [reply]
Only relevant to v4. V3 do not use plugins.

**Reply by iainhallam** [reply]
> Regarding server-side validation. You are right. Currently, you can connect to a custom API if that is a requirement. However, we have some exciting ideas on how this will be implemented in Lowdefy in the future.

Does this mean Lowdefy doesn't currently secure its back end in case a user of an app bypasses the front end to send unauthorised requests?

**Reply by Gervwyk** [reply]
@iainhallam NO - NOT AT ALL. Authentication and authorization is checked on all requests. 

We just do not validate the request data payload on the server. But this is something you can build into the request method on v4 if essential by writing a custom request plugin.

**Reply by carwei** [reply]
> > Regarding server-side validation. You are right. Currently, you can connect to a custom API if that is a requirement. However, we have some exciting ideas on how this will be implemented in Lowdefy in the future.
> 
> Does this mean Lowdefy doesn't currently secure its back end in case a user of an app bypasses the front end to send unauthorised requests?

Based on the information I gathered so far, the user needs a signed token from SSO that holds their roles in the claims. So unless they forged the token they only have the roles they should have. But they could in theory send anything to any end-point they have access to with that role, bypassing the front end to send something unintended, which could be a problem in its own. On an internal app the risk could be seen as manageable, depending on how sensitive the data is of course. But if somebody does something (and get caught), at least you know exactly who that person is (not a random person on the internet).

Would that be a correct description @Gervwyk?

**Reply by Gervwyk** [reply]
Yes, correct, user roles are part of the signed token. Tokens are validated on every request. In Lowdefy request, config are never shared with the client, only id - so user are not aware of the request logic and details. Requests are linked to pages. 

Pages can be public, private and protected by roles. 

Requests on public pages can be fired by anyone from the same domain, (@SamTolmay please check) since  public pages do not include auth. 

Private pages are authed. When pages are protected by roles, a user can trigger the requests on the page grated that they pass auth and that they have the applied role in the token claim. So, in short, if a user has access to a page, they can fire the requests on the page. For more strict validation, server-side logic can be defined in config, for example setting the filter object in a mongodb update as an additional check. 

Consider:
```yaml
id: update_my_post
type: MongoDBUpdateOne
connectionId: posts
payload:
  _id:
    _state: post._id
  message:
    _state: post.message
properties:
  filter:
    _id:
      _payload: _id
    created_by:
      _user: id
  update:
    $set:
      message:
        _payload: message
```
NOTE: Do not pass the user object in payload on v4. It is available and validated via the signed token on the server. In v3 the same applies, but the entire state is passed to the server on request, not just the payload.

**Reply by iainhallam** [reply]
> @iainhallam NO - NOT AT ALL. Authentication and authorization is checked on all requests.
> 
> We just do not validate the request data payload on the server. But this is something you can build into the request method on v4 if essential by writing a custom request plugin.

Ah, OK; thanks for that.
